File: packages/android-native/app/src/main/java/dev/tatliving/palmvellum/organizers/data/model/Fields.kt
package dev.tatliving.palmvellum.organizers.data.model

import kotlinx.serialization.Serializable
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

/** Shared JSON codec for record metadata + (later) sync payloads. */
val PalmJson = Json {
    ignoreUnknownKeys = true
    encodeDefaults = true
}

/**
 * To Do metadata (records.type='todo'). Key names match the PWA's
 * `metadata.palm_*` so cloud sync stays compatible.
 */
@Serializable
data class TodoFields(
    val palm_due_date: String? = null,   // yyyy-MM-dd
    val palm_priority: Int? = null,        // 1..5
    val palm_completed: Boolean = false,
    val palm_notes: String? = null,
    val palm_category_name: String? = null,
)

/**
 * Address metadata (records.type='contact'). P1 keeps a single phone +
 * email for simplicity; the PWA uses a phones[] array — P2 sync mapping
 * will reconcile.
 */
@Serializable
data class ContactFields(
    val palm_first_name: String? = null,
    val palm_last_name: String? = null,
    val palm_company: String? = null,
    val palm_title: String? = null,
    val palm_phone: String? = null,
    val palm_email: String? = null,
    val palm_notes: String? = null,
)

/**
 * Expense metadata (records.type='expense'). Vendor doubles as records.body.
 * Key names match the PWA's `metadata.palm_*` so cloud sync stays compatible.
 */
@Serializable
data class ExpenseFields(
    val palm_amount: Double? = null,
    val palm_currency: String? = null,
    val palm_vendor: String? = null,
    val palm_expense_type: String? = null,
    val palm_payment: String? = null,
    val palm_expense_date: String? = null, // yyyy-MM-dd
    val palm_city: String? = null,
    val palm_attendees: String? = null,
    val palm_notes: String? = null,
    val palm_category_name: String? = null,
)

/**
 * Mail record metadata (records.type='mail'). The digest text lives in
 * records.body; these mirror the PWA's `metadata.mail_*` keys.
 */
@Serializable
data class MailFields(
    val mail_subject: String? = null,
    val mail_from: String? = null,
    val mail_source_name: String? = null,
    val mail_source_url: String? = null,
    val mail_source_type: String? = null,
    val mail_topic: String? = null,
    val mail_references: List<String>? = null,
    val mail_date_local: String? = null,
    val mail_fetched_at: String? = null,
)

/**
 * A row of the `mail_sources` table (subscriptions). Not a record — synced
 * directly via PostgREST, not through the local Room mirror.
 */
@Serializable
data class MailSource(
    val id: String,
    val user_id: String? = null,
    val name: String,
    val source_type: String = "url",
    val url: String? = null,
    val topic: String? = null,
    val fetch_time: String = "07:00:00",
    val timezone: String = "UTC",
    val enabled: Boolean = true,
    val digest_hint: String? = null,
    val output_language: String? = null,
)

fun TodoFields.toJson(): String = PalmJson.encodeToString(this)
fun ContactFields.toJson(): String = PalmJson.encodeToString(this)
fun ExpenseFields.toJson(): String = PalmJson.encodeToString(this)

fun todoFieldsFrom(json: String): TodoFields =
    runCatching { PalmJson.decodeFromString<TodoFields>(json) }.getOrDefault(TodoFields())

fun contactFieldsFrom(json: String): ContactFields =
    runCatching { PalmJson.decodeFromString<ContactFields>(json) }.getOrDefault(ContactFields())

fun expenseFieldsFrom(json: String): ExpenseFields =
    runCatching { PalmJson.decodeFromString<ExpenseFields>(json) }.getOrDefault(ExpenseFields())

fun mailFieldsFrom(json: String): MailFields =
    runCatching { PalmJson.decodeFromString<MailFields>(json) }.getOrDefault(MailFields())

/** A To Do task the Memo AI suggested; the user approves it before it's created. */
@Serializable
data class ProposedTodo(
    val description: String = "",
    val due_date: String? = null,
    val priority: Int? = null,
    val notes: String? = null,
)

/** Memo AI bookkeeping kept in a thought record's metadata. */
@Serializable
data class MemoAiFields(
    val proposed_todos: List<ProposedTodo> = emptyList(),
    val added_todos: List<String> = emptyList(),
)

fun memoAiFrom(json: String): MemoAiFields =
    runCatching { PalmJson.decodeFromString<MemoAiFields>(json) }.getOrDefault(MemoAiFields())

File: packages/mac-native/Sources/PalmKit/Fields.swift
import Foundation

// Type-specific structured fields live in the record's `metadata` JSON blob
// under `palm_*` / `mail_*` keys, exactly matching the PWA + Android so the
// three clients interoperate. `body` carries the free text.

// MARK: - JSON helpers

public enum PalmJSON {
    public static func dict(_ json: String) -> [String: Any] {
        guard let data = json.data(using: .utf8),
              let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        else { return [:] }
        return obj
    }

    public static func string(_ dict: [String: Any]) -> String {
        guard let data = try? JSONSerialization.data(withJSONObject: dict, options: [.sortedKeys]),
              let s = String(data: data, encoding: .utf8)
        else { return "{}" }
        return s
    }
}

extension Encodable {
    /// Encode self into a `[String: Any]` (using the type's CodingKeys → palm_* names).
    func asDict() -> [String: Any] {
        guard let data = try? JSONEncoder().encode(self),
              let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        else { return [:] }
        return obj
    }

    /// Overlay self onto an existing metadata JSON string, preserving any
    /// keys we don't manage (e.g. AI / upload metadata).
    public func merged(into existingJSON: String) -> String {
        var base = PalmJSON.dict(existingJSON)
        for (k, v) in asDict() { base[k] = v }
        return PalmJSON.string(base)
    }
}

// MARK: - Todo

public struct TodoFields: Codable, Equatable {
    public var palmDueDate: String = ""        // yyyy-MM-dd or ""
    public var palmPriority: Int = 3           // 1..5
    public var palmCompleted: Bool = false
    public var palmNotes: String = ""
    public var palmCategoryName: String = "Unfiled"

    enum CodingKeys: String, CodingKey {
        case palmDueDate = "palm_due_date"
        case palmPriority = "palm_priority"
        case palmCompleted = "palm_completed"
        case palmNotes = "palm_notes"
        case palmCategoryName = "palm_category_name"
    }

    public init() {}

    public init(from json: String) {
        let d = PalmJSON.dict(json)
        palmDueDate = d["palm_due_date"] as? String ?? ""
        palmPriority = (d["palm_priority"] as? Int) ?? Int(d["palm_priority"] as? Double ?? 3)
        palmCompleted = d["palm_completed"] as? Bool ?? false
        palmNotes = d["palm_notes"] as? String ?? ""
        palmCategoryName = d["palm_category_name"] as? String ?? "Unfiled"
    }
}

// MARK: - Contact

public struct ContactFields: Codable, Equatable {
    public var palmFirstName: String = ""
    public var palmLastName: String = ""
    public var palmCompany: String = ""
    public var palmTitle: String = ""
    public var palmPhone: String = ""
    public var palmEmail: String = ""
    public var palmNotes: String = ""

    enum CodingKeys: String, CodingKey {
        case palmFirstName = "palm_first_name"
        case palmLastName = "palm_last_name"
        case palmCompany = "palm_company"
        case palmTitle = "palm_title"
        case palmPhone = "palm_phone"
        case palmEmail = "palm_email"
        case palmNotes = "palm_notes"
    }

    public init() {}

    public init(from json: String) {
        let d = PalmJSON.dict(json)
        palmFirstName = d["palm_first_name"] as? String ?? ""
        palmLastName = d["palm_last_name"] as? String ?? ""
        palmCompany = d["palm_company"] as? String ?? ""
        palmTitle = d["palm_title"] as? String ?? ""
        palmPhone = d["palm_phone"] as? String ?? ""
        palmEmail = d["palm_email"] as? String ?? ""
        palmNotes = d["palm_notes"] as? String ?? ""
    }

    public var displayName: String {
        let last = palmLastName.trimmingCharacters(in: .whitespaces)
        let first = palmFirstName.trimmingCharacters(in: .whitespaces)
        if !last.isEmpty && !first.isEmpty { return "\(last), \(first)" }
        if !last.isEmpty { return last }
        if !first.isEmpty { return first }
        if !palmCompany.isEmpty { return palmCompany }
        return "(no name)"
    }
}

// MARK: - Expense

public struct ExpenseFields: Codable, Equatable {
    public var palmAmount: Double = 0
    public var palmCurrency: String = "HKD"
    public var palmVendor: String = ""
    public var palmExpenseType: String = "Unfiled"
    public var palmPayment: String = "Unfiled"
    public var palmExpenseDate: String = ""
    public var palmCity: String = ""
    public var palmAttendees: String = ""
    public var palmNotes: String = ""
    public var palmCategoryName: String = "Unfiled"

    enum CodingKeys: String, CodingKey {
        case palmAmount = "palm_amount"
        case palmCurrency = "palm_currency"
        case palmVendor = "palm_vendor"
        case palmExpenseType = "palm_expense_type"
        case palmPayment = "palm_payment"
        case palmExpenseDate = "palm_expense_date"
        case palmCity = "palm_city"
        case palmAttendees = "palm_attendees"
        case palmNotes = "palm_notes"
        case palmCategoryName = "palm_category_name"
    }

    public init() {}

    public init(from json: String) {
        let d = PalmJSON.dict(json)
        palmAmount = (d["palm_amount"] as? Double) ?? Double(d["palm_amount"] as? Int ?? 0)
        palmCurrency = d["palm_currency"] as? String ?? "HKD"
        palmVendor = d["palm_vendor"] as? String ?? ""
        palmExpenseType = d["palm_expense_type"] as? String ?? "Unfiled"
        palmPayment = d["palm_payment"] as? String ?? "Unfiled"
        palmExpenseDate = d["palm_expense_date"] as? String ?? ""
        palmCity = d["palm_city"] as? String ?? ""
        palmAttendees = d["palm_attendees"] as? String ?? ""
        palmNotes = d["palm_notes"] as? String ?? ""
        palmCategoryName = d["palm_category_name"] as? String ?? "Unfiled"
    }
}
